Change units to astropy units / quantities

dinossimpson opened the pull request:
This is the latest version of the unit_refactoring branch and it includes the change of center_frequency to a Quantity.

keflavich commented on pyspeckit/spectrum/fitters.py:
```diff
@@ -1109,10 +1109,19 @@ def annotate(self,loc='upper right',labelspacing=0.25, markerscale=0.01,
+                unit_key = 'micron'
+            elif unit_key == 'angstroms':
+                unit_key = 'Angstrom'
+            xchar = xcharconv[u.Unit(unit_key).physical_type]
```
We might be able to use the `unit.to_string` method for this in the future

keflavich commented on pyspeckit/spectrum/models/hyperfine.py:
```diff
@@ -257,7 +258,7 @@ def hyperfine(self, xarr, Tex=5.0, tau=0.1, xoff_v=0.0, width=1.0,
                     tau_line = (tau * self.line_strength_dict[linename]/self.relative_strength_total_degeneracy[linename])
           
-                tau_nu = np.array(tau_line * np.exp(-(xarr+nuoff-self.freq_dict[linename])**2/(2.0*nuwidth**2)))
+                tau_nu = np.array(tau_line * np.exp(-(xarr.value+nuoff-self.freq_dict[linename])**2/(2.0*nuwidth**2)))
```
instead of doing `.value` here we can do it on line 212 above

keflavich commented on pyspeckit/spectrum/models/model.py:
```diff
@@ -307,18 +319,6 @@ def f(p,fjac=None):
                 return [0,residuals]
         return f
 
-    def __call__(self, *args, **kwargs):
```
The `__call__` method should remain; there is no reason to remove it.  Just remove the `multisingle` parsing part of it.

dinossimpson replied:
this is just moved this below the __init__ method, and in the multisingle branch the check for multisingle is gone, should I rebase on top of it?

keflavich replied:
oh.... this is OK.  I just missed that.

keflavich commented on pyspeckit/spectrum/plotters.py:
```diff
@@ -39,6 +40,11 @@ def __init__(self, Spectrum, autorefresh=True, title="", ylabel="",
+        if self._xunit == 'microns':
+            self._xunit = 'micron'
+        elif self._xunit == 'angstroms':
+            self._xunit = 'Angstrom'
```
Where are the plural versions of these units coming from?  They should just be replaced with their singular variants wherever needed.  If they really are necessary, we should use a universal `plural_units_mapping` dictionary {'angstroms':'angstrom'}  somewhere (in `units.py`?) instead of using two `if` statements.

keflavich commented on pyspeckit/spectrum/units.py:
```diff
@@ -958,26 +667,38 @@ def _make_header(self, tolerance=1e-8):
+        """
+        # equivalency finding
+        if velocity_convention:
+            print("generating the equivalencies")
```
need to remove print statements (or if they're really necessary, use `log.info`, where `log` is `astropy.log`)

keflavich commented on pyspeckit/spectrum/writers/fits_writer.py:
```diff
@@ -39,9 +39,17 @@ def write_data(self, filename=None, newsuffix='out', clobber=True,
+                        try:
+                            header[k] = v.value
+                        except AttributeError:
+                            pass
```
Why `pass` here?  If the header is not settable, I think we might want to raise an exception

keflavich commented:
@dinossimpson - looking great.  There are some minor changes and cleanup, and a few places where exceptions need to be treated more conservatively, but this is nearly there.

keflavich commented:
@dinossimpson let me know when you have the travis tests passing again *or* if you need me to do something additional to get travis to work (e.g., pull & push your branch of `pyspeckit-tests`).  Do so by adding a comment in this Issue so that I get a notification.

dinossimpson commented:
@keflavich this branch should pass the tests I pushed in cfquantity in pyspeckit-tests and it includes the changes from the review.

keflavich commented:
OK, good - I restarted the build.

keflavich commented:
It turns out those tex failures are also happening on travis.  We need to investigate them further.

dinossimpson commented:
@keflavich yes indeed, I thought it was only a local error. I'll try to see which line causes it

keflavich commented:
pushed; build restarted

keflavich commented:
You seem to have encountered an infinite recursion in unit conversion.  There also seems to be some location where the string version of a unit is being treated as a unit object (https://travis-ci.org/pyspeckit/pyspeckit/jobs/55482248#L1647)

keflavich commented:
Strangely, this only affects np 1.6 and np 1.7  - it may be OK to remove those as long as we add np 1.9.   But this doesn't solve both issues

dinossimpson commented:
Oh ok, I'll try again to reproduce it

keflavich commented:
@dinossimpson this needs to be rebased so it can be merged.

dinossimpson commented:
@keflavich Ok! I'll rebase on top of master, I still can't reproduce the error happening in #323.3 with ASTROPY_VERSION=devel though.

keflavich commented:
restarted build

keflavich commented:
Good!  Green except for [`AttributeError: 'str' object has no attribute 'decompose'`](https://travis-ci.org/pyspeckit/pyspeckit/jobs/56897880#L1688).  Close!

dinossimpson commented:
Hey @keflavich, with the last push the Quantity initialization bug should be fixed ( the AttributeError: 'str' object has no attribute 'decompose'. one), also I added a __init__.py in the examples so that they are included in the build, and added the ammonia_fit_example.py in run_tests.py. I also added a commented out run_examples.py if we decide on running the tests from a different file than the one with the tests (this requires a solution to a small importing problem). Finally, I am missing some test files needed in the examples to run them and I am not sure where to find them.

dinossimpson commented:
more specifically these 3 are missing: (with their corresponding examples)
gbt_sdfits_example.py: AGBT09C_049_02.raw.acs.fits
h2co_mm_example.py: 303-202_321-220_5kms_temperature_para_tex1.fits
n2hp_example.py: n2hp_opha_example.fits

keflavich commented:
OK, we can't run the `gbt_sdfits_example` because I don't have that file and, in any case, it would be too big.

Same for the `h2co_mm_example`: the file is too big.

For n2hp_ohpa_example, it's here: https://github.com/pyspeckit/pyspeckit/blob/master/examples/n2hp_opha_example.fits

keflavich commented:
You need to merge or rebase against master

keflavich commented on examples/agn_example.py:
```diff
@@ -117,7 +118,7 @@
 for i, line in enumerate(spec.measurements.lines.keys()):
     x = spec.measurements.lines[line]['modelpars'][1]
-    spec.plotter.axis.plot([x]*2, [spec.plotter.ymin, spec.plotter.ymax], ls = '--', color = 'k')
+    spec.plotter.axis.plot([x]*2, [spec.plotter.ymin.value, spec.plotter.ymax.value], ls = '--', color = 'k')
```
Does matplotlib still fail if you leave out the `.value`?  It shouldn't... I'm concerned.

dinossimpson replied:
yes I didn't look into it but I got this error here:
ValueError: setting an array element with a sequence.

keflavich replied:
hmm... yes, that's a problem.  can do `u.Quantity([spec.plotter.ymin.value, spec.plotter.ymax.value])` but I'm not sure that's better... darn.

keflavich commented on examples/ammonia_fit_example.py:
```diff
@@ -1,7 +1,8 @@
 import pyspeckit
+import astropy.units as u
```
is this import being used?

dinossimpson replied:
removed

keflavich commented on examples/ammonia_fit_example.py:
```diff
@@ -1,7 +1,8 @@
 
 # Grab a .fits spectrum with a legitimate header
-sp = pyspeckit.Spectrum('../pyspeckit/tests/G031.947+00.076_nh3_11_Tastar.fits')
+sp = pyspeckit.Spectrum('G031.947+00.076_nh3_11_Tastar.fits')
```
It looks like you've fixed all of the examples to work in the appropriate path?  That's good, I'm just commenting that I've seen it.  The relative paths that you're removing are things I had to add over the weekend.

dinossimpson replied:
Yes but it is still sketchy. These paths work if the examples are ran through python setup.py test, if a user tries to run a single example the paths break.

keflavich replied:
Yeah, that's annoying.  Perhaps it would be best to include symbolic links in the `examples/` directory

keflavich commented on examples/interactive_example_hr2421.py:
```diff
@@ -32,11 +33,11 @@
-
+sp.plotter.savefig('interactive_example_hr2421_baseline_firstclick.png', bbox_inches=None)
 event2 = matplotlib.backend_bases.MouseEvent('button_press_event', sp.plotter.axis.figure.canvas,850,280,button=1)
+# event2.xdata and .ydata is None here
```
I can reproduce this.  I don't understand it, though.  This is frustrating and we'll have to return to it.... for now, I guess, set it manually:
`event2.xdata = 850`

keflavich commented on examples/n2hp_cube_example.py:
```diff
@@ -23,7 +23,7 @@
 
 # Get a measurement of the error per pixel
-errmap = spc.slice(20, 28, unit='km/s').cube.std(axis=0)
+errmap = spc.slice(20, 28, units='km/s').cube.std(axis=0)
```
We can allow `unit` or `units`, I think, to avoid having to make this sort of change.  But this is an aside, low-priority...

dinossimpson commented:
hmm I can't reproduce this locally: https://travis-ci.org/pyspeckit/pyspeckit/jobs/57795387
it fails at the agn_example.py:130 because something has a unit when it shouldn't and it must be converted to a float, but I don't understand why it wouldn't fail on my machine.

keflavich commented:
re: https://github.com/pyspeckit/pyspeckit/pull/26#issuecomment-91231960 - any chance you're using a different astropy version?  If not, yeah, hard to say what's going on.

keflavich commented on pyspeckit/spectrum/writers/fits_writer.py:
```diff
@@ -41,7 +41,16 @@ def write_data(self, filename=None, newsuffix='out', clobber=True,
+                        header[k] = v
+                    except ValueError:
+                        try:
+                            #v is a SpectroscopicAxis
```
that shouldn't be - header keywords have to be scalar.  It could be that `v` is a `Quantity`, though

dinossimpson replied:
yes indeed. changed

keflavich commented on pyspeckit/spectrum/units.py:
```diff
@@ -983,26 +668,37 @@ def _make_header(self, tolerance=1e-8):
+                           center_frequency=None, equivalencies=[]):
         """
+        Utility function to add equivalencies from the velocity_convention 
+        and the center_frequency
```
Since this function is exposed to users, can you add more documentation?  Specifically, 
```
Parameters
------------
velocity_convention : str
```
etc?

keflavich commented on pyspeckit/spectrum/units.py:
```diff
@@ -665,31 +512,27 @@ def convert_to_unit(self, unit, **kwargs):
+        try:
+            unit = u.Unit(unit)
+        except ValueError:
+            raise ValueError('Invalid unit')
```
Won't this raise a `u.UnitsError`?  If not, maybe it can.   This is minor though.

dinossimpson replied:
changed it to validate_unit

keflavich commented on pyspeckit/spectrum/units.py:
```diff
@@ -453,83 +395,64 @@ def __array_wrap__(self,out_arr,context=None):
+        try:
+            if unit is None or unit == 'unknown':
+                unit = u.dimensionless_unscaled
+            if unit == 'angstroms': unit = 'angstrom'
```
we probably want a lookuptable for this but for now it's fine

keflavich commented on pyspeckit/spectrum/units.py:
```diff
@@ -453,83 +395,64 @@ def __array_wrap__(self,out_arr,context=None):
+            if unit == 'angstroms': unit = 'angstrom'
+            unit = u.Unit(unit)
+        except ValueError:
+            if bad_unit_response=="pixel":
```
where is `bad_unit_response` defined?  Is it a global above?

keflavich replied:
I think that needs to be passed to `validate_unit` as a parameter, maybe?

dinossimpson replied:
yes of course I forgot it, added

keflavich commented on pyspeckit/cubes/SpectralCube.py:
```diff
@@ -4,12 +4,18 @@
-   * `pyregion <git://github.com/astropy/pyregion.git>`_
+   * `coords <http://stsdas.stsci.edu/astrolib/coords-0.37.tar.gz>`_ (`homepage <http://www.scipy.org/AstroLibCoordsHome>`_)
+   * `pyregion <git://github.com/leejjoon/pyregion.git>`_
+   * `pywcs <git://github.com/astropy/astropy.git>`_
```
These lines were modified in a recent commit and should not be reverted.

keflavich commented on pyspeckit/cubes/SpectralCube.py:
```diff
@@ -4,12 +4,18 @@
 
+In the near future, the coords_ requirement will be replaced with astropy_\'s
+coordinates module.
+    
```
this should be gone

keflavich commented on pyspeckit/cubes/SpectralCube.py:
```diff
@@ -31,78 +37,39 @@
+        Alternatively, you can specify the *xarr*, *cube*, and *header* kwargs.
+        If nothing is specified, a blank :Cube: will be generated.
 
+        x0,y0 - initial spectrum to use (defaults to lower-left corner)
```
None of the changes should be reverted here (right now, your PR is reverting changes I made)

keflavich commented on pyspeckit/cubes/SpectralCube.py:
```diff
@@ -218,27 +165,23 @@ def copy(self,deep=True):
+        *stop*  [ numpy.float or int ]
             stop of slice
-        unit : str
+        *units* [ str ]
```
The `unit : str` format is correct, `*start*` is incorrect

keflavich commented on examples/agn_example.py:
```diff
@@ -26,9 +27,9 @@
-spec.xarr.units='angstroms'
+spec = pyspeckit.Spectrum('sample_sdss.txt')
+spec.unit = 'erg s^{-1} cm^{-2} \\AA^{-1}'
+spec.xarr.set_unit = u.dimensionless_unscaled
```
`set_unit` is a method.  You should be setting the unit to `angstroms` (or `u.AA`) here anyway:
`spec.xarr.set_unit(u.AA)`

dinossimpson replied:
WELL, this is embarrassing. I wasn't even drunk.

keflavich commented on examples/hcn_example.py:
```diff
@@ -1,11 +1,13 @@
-sp.xarr.units='km/s'
-sp.xarr.refX = 88.63184666e9
+sp = pyspeckit.Spectrum('02232_plus_6138.txt')
+sp.xarr._unit=u.km/u.s
```
`set_unit` here

dinossimpson replied:
fixed

keflavich commented on examples/run_examples.py:
```diff
@@ -1,9 +1,25 @@
+# interactive=False
+# import subprocess
+# import os
+# import matplotlib
```
Could you add some explanation for why the whole file is commented out now?  Perhaps a `""" docstring """` at the top indicating what replaced it?

dinossimpson replied:
sure, I'll say this is a dummy run_examples file that will be used in a later stage

keflavich commented on pyspeckit/spectrum/classes.py:
```diff
@@ -181,14 +180,12 @@ def __init__(self, filename=None, filetype=None, xarr=None, data=None,
 
-        # it is very important that this be done BEFORE the spectofit is set!
-        self._sort()
-
```
please note this comment - this is one of the cases where the merge went the wrong way.  There will be hard-to-understand crashes if `_sort` is done after `specfit` is initialized.

dinossimpson replied:
I see, the merge errors worry me a bit, I'll try to find the commit in your branch where it happened and make sure I didn't skip anything. I believe the error was caused by me using git checkout --ours and --theirs

keflavich commented on pyspeckit/spectrum/classes.py:
```diff
@@ -291,17 +288,22 @@ def parse_text_header(self,Table):
-            self.xarr.unit = Table.columns[xtype].unit
+            unit = Table.columns[xtype].unit
+            if unit == 'angstroms' : unit = 'angstrom'
+            self.xarr._unit = u.Unit(unit)
```
should use `set_unit` here, and `set_unit` should handle the angstrom/angstroms issue

dinossimpson replied:
fixed

keflavich commented on pyspeckit/spectrum/classes.py:
```diff
@@ -291,17 +288,22 @@ def parse_text_header(self,Table):
-            self.xarr.unit = Table.columns[xtype].unit
+            unit = Table.columns[xtype].unit
+            if unit == 'angstroms' : unit = 'angstrom'
+            self.xarr._unit = u.Unit(unit)
```
set_unit

dinossimpson replied:
fixed

keflavich commented on pyspeckit/spectrum/classes.py:
```diff
@@ -291,17 +288,22 @@ def parse_text_header(self,Table):
             warn( "Warning: Invalid xtype in text header - this may mean no text header was available.  X-axis units will be pixels unless you set them manually (e.g., sp.xarr.unit='angstroms')")
             self.xarr.xtype = 'pixels'
-            self.xarr.units = 'none'
+            self.xarr._unit = u.pixel
```
set_unit

dinossimpson replied:
fixed

keflavich commented on pyspeckit/spectrum/classes.py:
```diff
@@ -328,11 +330,11 @@ def parse_hdf5_header(self, hdr):
         
         self.xarr.xtype = hdr['xtype']
-        self.xarr.xunit = hdr['xunit']
+        self.xarr.xunits = hdr['xunits']
```
why the added 's'?  I think this is a bad merge.

keflavich commented on pyspeckit/spectrum/classes.py:
```diff
@@ -375,7 +377,7 @@ def measure(self, z=None, d=None, fluxnorm=None, miscline=None,
                 ignore=ignore, derive=derive, **kwargs)
 
-    def crop(self, x1, x2, unit=None, **kwargs):
+    def crop(self, x1, x2, units=None, **kwargs):
```
do not add plural; likely bad merge

dinossimpson replied:
yes probably, changed it to 'unit' in this file

keflavich commented on pyspeckit/spectrum/fitters.py:
```diff
@@ -1764,9 +1768,12 @@ def measure_approximate_fwhm(self, threshold='error', emission=True,
             newxarr = units.SpectroscopicAxis(
-                    np.arange(xarr.min()-cd,xarr.max()+cd,cd / float(interpolate_factor)))
+                    np.arange(xarr.min().value-cd,xarr.max().value+cd,cd / float(interpolate_factor)),
+                    unit=xarr._unit,
```
just `.unit`

dinossimpson replied:
fixed

keflavich commented on pyspeckit/spectrum/models/hyperfine.py:
```diff
@@ -121,28 +122,26 @@ def __call__(self, *args, **kwargs):
-                      return_hyperfine_components=False, Tbackground=2.73,
-                      Tex=5.0, tau=0.1):
+    def hyperfine_amp(self, xarr, amp=None, xoff_v=0.0, width=1.0, 
+            return_hyperfine_components=False, Tbackground=2.73, Tex=5.0, tau=0.1):
```
bad merge

dinossimpson replied:
fixed

keflavich commented on pyspeckit/spectrum/models/hyperfine.py:
```diff
@@ -210,7 +209,7 @@ def hyperfine(self, xarr, Tex=5.0, tau=0.1, xoff_v=0.0, width=1.0,
 
         # Convert X-units to frequency in Hz
-        xarr = xarr.as_unit('Hz')
+        xarr = xarr.as_unit('Hz', equivalencies=u.doppler_radio(xarr.center_frequency)).value
```
looks good!

keflavich commented on pyspeckit/spectrum/models/hyperfine.py:
```diff
@@ -257,12 +256,9 @@ def hyperfine(self, xarr, Tex=5.0, tau=0.1, xoff_v=0.0, width=1.0,
                     # the total optical depth, which is being fitted, should be the sum of the components
-                    tau_line = (tau * self.line_strength_dict[linename] /
-                                self.relative_strength_total_degeneracy[linename])
+                    tau_line = (tau * self.line_strength_dict[linename]/self.relative_strength_total_degeneracy[linename])
```
probably bad merge

dinossimpson replied:
fixed

keflavich commented on pyspeckit/spectrum/models/test_moments.py:
```diff
@@ -0,0 +1,26 @@
+		# if moments() returns wrong number of parameters this should fail 
+		print model_instance.n_modelfunc(pars=params)(xarr)
+	except Exception as e:
+		print e
```
This function looks great, but you could (maybe should) put it into a different pull request, since it's not directly related to the Quantity refactor

keflavich commented on pyspeckit/spectrum/plotters.py:
```diff
@@ -320,11 +343,12 @@ def reset_limits(self,xmin=None, xmax=None, ymin=None, ymax=None,
-                if not self.silent and not reset_ylimits: warn( "Resetting Y-axis min/max because the plot is out of bounds." )
-                self.ymin = None
-                self.ymax = None
+            if self.ymin and self.ymax:
```
Is this needed?  It looks like it is, but I don't recall ever encountering an error here.  This is probably the right thing to do, I just want to know if I had missed a case previously

dinossimpson replied:
I made this change in one of the first commits but I remember it had to do with a NoneType error, I'll run the examples again without it to make sure.

keflavich commented on pyspeckit/spectrum/plotters.py:
```diff
@@ -341,19 +365,19 @@ def reset_limits(self,xmin=None, xmax=None, ymin=None, ymax=None,
-            self.ymin += self.offset
-            self.ymax += self.offset
+            self.ymin += u.Quantity(self.offset, self.ymin.unit)
+            self.ymax += u.Quantity(self.offset, u.Unit(self.ymax.unit))
```
These lines should look more similar; I think 376 is correct and 377 is unnecessary.

dinossimpson replied:
fixed

keflavich commented on pyspeckit/spectrum/plotters.py:
```diff
@@ -378,14 +402,13 @@ def label(self, title=None, xlabel=None, ylabel=None, verbose_label=False,
             if verbose_label:
                 self.xlabel = "%s %s %s" % ( self.Spectrum.xarr.velocity_convention.title(),
-                        self.Spectrum.xarr.frame.title(),
+                        # self.Spectrum.xarr.frame.title(),
```
If you remove this, you need to remove one of the `%s` formatting strings.  Is there any way to recover the `frame` from the unit?  I think the answer is "no", so removing it is the right thing to do, at least for now.

(physics lesson)
The `frame` means the rest frame of the data.  So for example, there is the "topocentric" frame, which means the frame of the observatory.  "geocentric" means the Earth's center of mass frame - "topocentric" is moving at up to 0.5 km/s relative to "geocentric" because of the Earth's rotation.  Then there are "heliocentric" (sun-centered) and "barycentric" (around the center of mass of our solar system, which is tilted a little bit toward Jupiter), and then "Local Standard of Rest", which is defined by the average motions of stars near Earth.

dinossimpson replied:
removed, wow! there are so many factors to take into account! Also, it is the first time I realise the solar system has a center of mass :open_mouth:

keflavich commented on pyspeckit/spectrum/plotters.py:
```diff
@@ -378,14 +402,13 @@ def label(self, title=None, xlabel=None, ylabel=None, verbose_label=False,
-        elif isinstance(self.Spectrum.xarr.xtype,str):
-            self.xlabel = self.Spectrum.xarr.xtype.title()
+        elif self._xunit:
+            self.xlabel += " ("+u.Unit(self._xunit).to_string()+")"
```
This will only show `(km/s)`, for example.  We want something like "Velocity (km/s)".  This can be achieved with something like:
`"{0} ({1})".format(unit.physical_type.title(), unit.to_string())`
However, the physical type `speed` will show up where we want `Velocity`, so we need a lookuptable here.  Probably just a simple one:
`{'speed':'velocity'}` and the rest would just use their defaults.

keflavich replied:
(`xtype` was one of  `velocity, wavelength, frequency, pixels`)

dinossimpson replied:
added the dictionary on top of the file

keflavich commented on pyspeckit/spectrum/units.py:
```diff
@@ -354,80 +315,63 @@ def __new__(self, xarr, unit="Hz", frame='rest', frame_offset=0.0,
             else:
-                subarr.refX_units = 'Hz'
+                refX_units = 'Hz'
+            subarr.refX_units = refX_units
```
should make this `refX_unit`

keflavich replied:
but we should really allow the plural version and start putting up deprecation warnings when it's used

dinossimpson replied:
fixed

keflavich commented on pyspeckit/spectrum/units.py:
```diff
@@ -453,83 +400,64 @@ def __array_wrap__(self,out_arr,context=None):
+            unit = u.Unit(unit)
+        except ValueError:
+            if bad_unit_response == "pixel":
+                unit = "unknown"
```
should this just be `u.dimensionless_unscaled` since you're setting it that way on line 426?

dinossimpson replied:
changed

dinossimpson commented on pyspeckit/spectrum/models/hyperfine.py:
```diff
@@ -257,12 +256,9 @@ def hyperfine(self, xarr, Tex=5.0, tau=0.1, xoff_v=0.0, width=1.0,
-                tau_nu = np.array(tau_line *
-                                  np.exp(-(xarr+nuoff-self.freq_dict[linename])**2 /
-                                         (2.0*nuwidth**2)))
+                tau_nu = np.array(tau_line * np.exp(-(xarr+nuoff-self.freq_dict[linename])**2/(2.0*nuwidth**2)))
```
fixed

dinossimpson commented:
Hey @keflavich, I did a proof read, left some comments, and I think it should be ready for review!

keflavich commented:
@dinossimpson - Looks good.  A few minor changes are warranted, but I consider this mergeable now.
